### ISSUE TITLE
After compiling *** [.pio\build\tronxy_stm32f446\firmware.elf] Error 1  - What means and how resolve it?

### ISSUE THREAD

**Issue by Aerospacesky** [tronxy3d/SIM480x320#17]
After compiling with Platformio the F4xx-SIM480x320, the final result is this below. I do not know how solve this issue. Can you help me please?

Usage: arm-none-eabi-objcopy [option(s)] in-file [out-file]
 Copies a binary file, possibly transforming it in the process
 The options are:
  -I --input-target <bfdname>      Assume input file is in format <bfdname>
  -O --output-target <bfdname>     Create an output file in format <bfdname>
  -B --binary-architecture <arch>  Set output arch, when input is arch-less
  -F --target <bfdname>            Set both input and output format to <bfdname>
     --debugging                   Convert debugging information, if possible
  -p --preserve-dates              Copy modified/access timestamps to the output
  -D --enable-deterministic-archives
                                   Produce deterministic output when stripping archives
  -U --disable-deterministic-archives
                                   Disable -D behavior (default)
  -j --only-section <name>         Only copy section <name> into the output
     --add-gnu-debuglink=<file>    Add section .gnu_debuglink linking to <file>
  -R --remove-section <name>       Remove section <name> from the output
     --remove-relocations <name>   Remove relocations from section <name>
  -S --strip-all                   Remove all symbol and relocation information
  -g --strip-debug                 Remove all debugging symbols & sections
     --strip-dwo                   Remove all DWO sections
     --strip-unneeded              Remove all symbols not needed by relocations
  -N --strip-symbol <name>         Do not copy symbol <name>
     --strip-unneeded-symbol <name>
                                   Do not copy symbol <name> unless needed by
                                     relocations
     --only-keep-debug             Strip everything but the debug information
     --extract-dwo                 Copy only DWO sections
     --extract-symbol              Remove section contents but keep symbols
  -K --keep-symbol <name>          Do not strip symbol <name>
     --keep-file-symbols           Do not strip file symbol(s)
     --localize-hidden             Turn all ELF hidden symbols into locals
  -L --localize-symbol <name>      Force symbol <name> to be marked as a local
     --globalize-symbol <name>     Force symbol <name> to be marked as a global
  -G --keep-global-symbol <name>   Localize all symbols except <name>
  -W --weaken-symbol <name>        Force symbol <name> to be marked as a weak
     --weaken                      Force all global symbols to be marked as weak
  -w --wildcard                    Permit wildcard in symbol comparison
  -x --discard-all                 Remove all non-global symbols
  -X --discard-locals              Remove any compiler-generated symbols
  -i --interleave[=<number>]       Only copy N out of every <number> bytes
     --interleave-width <number>   Set N for --interleave
  -b --byte <num>                  Select byte <num> in every interleaved block
     --gap-fill <val>              Fill gaps between sections with <val>
     --pad-to <addr>               Pad the last section up to address <addr>
     --set-start <addr>            Set the start address to <addr>
    {--change-start|--adjust-start} <incr>
                                   Add <incr> to the start address
    {--change-addresses|--adjust-vma} <incr>
                                   Add <incr> to LMA, VMA and start addresses
    {--change-section-address|--adjust-section-vma} <name>{=|+|-}<val>
                                   Change LMA and VMA of section <name> by <val>
     --change-section-lma <name>{=|+|-}<val>
                                   Change the LMA of section <name> by <val>
     --change-section-vma <name>{=|+|-}<val>
                                   Change the VMA of section <name> by <val>
    {--[no-]change-warnings|--[no-]adjust-warnings}
                                   Warn if a named section does not exist
     --set-section-flags <name>=<flags>
                                   Set section <name>'s properties to <flags>
     --set-section-alignment <name>=<align>
                                   Set section <name>'s alignment to <align> bytes
     --add-section <name>=<file>   Add section <name> found in <file> to output
     --update-section <name>=<file>
                                   Update contents of section <name> with
                                   contents found in <file>
     --dump-section <name>=<file>  Dump the contents of section <name> into <file>
     --rename-section <old>=<new>[,<flags>] Rename section <old> to <new>
     --long-section-names {enable|disable|keep}
                                   Handle long section names in Coff objects.
     --change-leading-char         Force output format's leading character style
     --remove-leading-char         Remove leading character from global symbols
     --reverse-bytes=<num>         Reverse <num> bytes at a time, in output sections with content
     --redefine-sym <old>=<new>    Redefine symbol name <old> to <new>
     --redefine-syms <file>        --redefine-sym for all symbol pairs
                                     listed in <file>
     --srec-len <number>           Restrict the length of generated Srecords
     --srec-forceS3                Restrict the type of generated Srecords to S3
     --strip-symbols <file>        -N for all symbols listed in <file>
     --strip-unneeded-symbols <file>
                                   --strip-unneeded-symbol for all symbols listed
                                     in <file>
     --keep-symbols <file>         -K for all symbols listed in <file>
     --localize-symbols <file>     -L for all symbols listed in <file>
     --globalize-symbols <file>    --globalize-symbol for all in <file>
     --keep-global-symbols <file>  -G for all symbols listed in <file>
     --weaken-symbols <file>       -W for all symbols listed in <file>
     --add-symbol <name>=[<section>:]<value>[,<flags>]  Add a symbol
     --alt-machine-code <index>    Use the target's <index>'th alternative machine
     --writable-text               Mark the output text as writable
     --readonly-text               Make the output text write protected
     --pure                        Mark the output file as demand paged
     --impure                      Mark the output file as impure
     --prefix-symbols <prefix>     Add <prefix> to start of every symbol name
     --prefix-sections <prefix>    Add <prefix> to start of every section name
     --prefix-alloc-sections <prefix>
                                   Add <prefix> to start of every allocatable
                                     section name
     --file-alignment <num>        Set PE file alignment to <num>
     --heap <reserve>[,<commit>]   Set PE reserve/commit heap to <reserve>/
                                   <commit>
     --image-base <address>        Set PE image base to <address>
     --section-alignment <num>     Set PE section alignment to <num>
     --stack <reserve>[,<commit>]  Set PE reserve/commit stack to <reserve>/
                                   <commit>
     --subsystem <name>[:<version>]
                                   Set PE subsystem to <name> [& <version>]
     --compress-debug-sections[={none|zlib|zlib-gnu|zlib-gabi}]
                                   Compress DWARF debug sections using zlib
     --decompress-debug-sections   Decompress DWARF debug sections using zlib
     --elf-stt-common=[yes|no]     Generate ELF common symbols with STT_COMMON
                                     type
     --verilog-data-width <number> Specifies data width, in bytes, for verilog output
  -M  --merge-notes                Remove redundant entries in note sections
      --no-merge-notes             Do not attempt to remove redundant notes (default)
  -v --verbose                     List all object files modified
  @<file>                          Read options from <file>
  -V --version                     Display this program's version number
  -h --help                        Display this output
     --info                        List object formats & architectures supported
arm-none-eabi-objcopy: supported targets: elf32-littlearm elf32-littlearm-fdpic elf32-bigarm elf32-bigarm-fdpic elf32-little elf32-big srec symbolsrec verilog tekhex binary ihex plugin        
*** [.pio\build\tronxy_stm32f446\firmware.elf] Error 1

**Comment by bytefix** [comment]
Edit: please use @Maker-Paul solution

~~@Aerospacesky This is an error in Tronxys own script, but it's not needed (I have not fully tested it yet, but it seems to work without as my tronxy still prints)~~
~~How to bypass the error:~~
~~1. in `ini\stm32f4.ini` delete `buildroot/share/PlatformIO/scripts/tronxy_scrypt.py` in line 659~~
~~2. compile (should work now)~~
~~3. copy the file `.pio\build\tronxy_stm32f446\firmware.bin` and paste it into the `update` folder~~
~~4. rename `firmware.bin` to `fmw_tronxy.bin`~~
~~5. copy the folder to your SD Card and boot your printer when the card is inserted~~

~~If you still get this error, try to remove the same script on line 476 in` ini\stm32f1.ini`~~

**Comment by Pelu88** [comment]
> @Aerospacesky This is an error in Tronxys own script, but it's not needed (I have not fully tested it yet, but it seems to work without as my tronxy still prints) How to bypass the error:
> 
> 1. in `ini\stm32f4.ini` delete `buildroot/share/PlatformIO/scripts/tronxy_scrypt.py` in line 659
> 2. compile (should work now)
> 3. copy the file `.pio\build\tronxy_stm32f446\firmware.bin` and paste it into the `update` folder
> 4. rename `firmware.bin` to `fmw_tronxy.bin`
> 5. copy the folder to your SD Card and boot your printer when the card is inserted
> 
> If you still get this error, try to remove the same script on line 476 in` ini\stm32f1.ini`

THIS ABSOLUTLY WORKED! Thanks a lot for share this solution

**Comment by Jcarter2141** [comment]
> @Aerospacesky This is an error in Tronxys own script, but it's not needed (I have not fully tested it yet, but it seems to work without as my tronxy still prints) How to bypass the error:
> 
> 1. in `ini\stm32f4.ini` delete `buildroot/share/PlatformIO/scripts/tronxy_scrypt.py` in line 659
> 2. compile (should work now)
> 3. copy the file `.pio\build\tronxy_stm32f446\firmware.bin` and paste it into the `update` folder
> 4. rename `firmware.bin` to `fmw_tronxy.bin`
> 5. copy the folder to your SD Card and boot your printer when the card is inserted
> 
> If you still get this error, try to remove the same script on line 476 in` ini\stm32f1.ini`

Tried this today on a new VEHO F446 board. I had no issues compiling the firmware but when I copy it to the update directory and rename it, it freezes at BURN:LOG_TRONXY.YPC.

**Comment by bytefix** [comment]
You need to define your project printer in /Marlin/TronxyMachine.h

**Comment by Jcarter2141** [comment]
> You need to define your project printer in /Marlin/TronxyMachine.h

I did that. Defined the machine as VEHO 800 2E. Everything is shaded except for the correct machine.
![20230823_141322](https://github.com/tronxy3d/F4xx-SIM480x320/assets/143022324/8fcf6e7f-846b-4f4d-ab7d-32819e29daf6)
![20230823_141349](https://github.com/tronxy3d/F4xx-SIM480x320/assets/143022324/fe524101-1d86-4971-b904-e6ee4a054a7f)

**Comment by Maker-Paul** [comment]
What ever you do DO NOT delete the lines in "ini\stm32f4.ini" or ini\stm32f1.ini that calls the "buildroot/share/PlatformIO/scripts/tronxy_scrypt.py" file this small script contains a post processing routine that copies the compiled output firmware file and removes a section from it and alters the offset so when you install the file it does not overwrite your boot loader. You may well be able to install the file the first time and it may well work but once the boot-loader is overwritten you will not be able to update by inserting an SD card.

The proper fix on my machine was to make sure no path contains a space in any directory or file name as the "buildroot/share/PlatformIO/scripts/tronxy_scrypt.py" output line is telling you it can't understand the instruction because probably the input path contains a space character so it can't resolve the file name. and therefor can't generate the proper adjusted output file

If you are seeing this :- 
     arm-none-eabi-objcopy [option(s)] in-file [out-file]
     Copies a binary file, possibly transforming it in the process 
    The options are:
          (followed by a load of instructions)
it means the instruction was not executed.


This is the instruction Scons (a post processing add on for Vs-code that no one mentioned before) uses. In the tronxy_scrypt.py file. 
            $OBJCOPY", "-O", "binary", "-R", ".eeprom"
This is failing. It requires two more parameters
            "$BUILD_DIR/${PROGNAME}.elf" (this provides the input file name)
and
            "tar_bin" (this provides the output file name)
The last parameter 
             "Building %s" % tar_bin"
simply reports what it is doing on your terminal output.
 
The join() in the env.Verbose sections just joins all three part together to construct the line of code that will be executed after compilation. which is :--- "arm-none-eabi-objcopy [option(s)] in-file [out-file]"

This is the contents of the tronxy_scrypt.py which does not and never has had a scripting mistake by tronxy.

import marlin
from SCons.Script import DefaultEnvironment

env = DefaultEnvironment()
board = env.BoardConfig()
if 'offset' in board.get("build").keys():
    marlin.relocate_vtab(board.get('build.offset'))

def output_target():
    tar_bin = "update/fmw_tronxy.bin"
    env.AddPostAction(
        "$BUILD_DIR/${PROGNAME}.elf",
        env.VerboseAction(" ".join([
            "$OBJCOPY", "-O", "binary", "-R", ".eeprom",
            "$BUILD_DIR/${PROGNAME}.elf", tar_bin
        ]), "Post Processing :-- %s" % tar_bin)
    )
output_target()

In conclusion Make sure your directory structure containing your VScode project does not include any directory names that include spaces or any other unusual characters as the  problem was probably caused by a rouge space in a directory name.

**Comment by bytefix** [comment]
@Maker-Paul 
Thank you for sharing your proper solution! I had no issues with the bootloader yet, so I did't look closer at the problem.

**Comment by Maker-Paul** [comment]
No problem took me a day to unscramble whats going on, as I had not come across Scons before. Which by the way I had to install on my Linux machine as it wasn't already installed as part of the VS-code install. 

I thought that would fix it so was a bit miffed when it didn't. I found a reference for Scons and worked out how it was being used as a wrapper by the python script to do something after compilation even though the python script executed before compilation (that makes your head hurt). That lead me to the "arm-none-eabi-objcopy [option(s)] in-file [out-file]" that was being executed after compilation and telling us it was having a problem, just not telling us what the problem was or even where this instruction was comming from. I had already put print statements into the python script to report the file paths so I new they were being constructed and used in this "arm-none-eabi-objcopy [option(s)] in-file [out-file]" command. Knowing my source file actually existed I was able to type the command into a terminal with the reported file paths and only then did it show the input file path had only been read up to the space in the name of the containing folder because it showed the command as seen by "arm-none-eabi-objcopy [option(s)] in-file [out-file]" in the terminal output with everything after the space missing. So it obviously couldn't resolve the input path file name.

This is more of a problem on Linux machines which have always been much more tolerant of folder name variation than windows. I don't have the facility to test the effect of not removing the piece of code or setting/not setting the offset. It may well depend on weather you update just by using the boot-loader or by using flashing software from a PC. But at least it will compile the way Tronxy intended.

**Comment by Maker-Paul** [comment]
> @Maker-Paul Thank you for sharing your proper solution! I had no issues with the bootloader yet, so I did't look closer at the problem.

No problem bytefix took me a day to unscramble whats going on, as I had not come across Scons before. Which by the way I had to install on my Linux machine as it wasn't already installed as part of the VS-code install. 

I thought that would fix it so was a bit miffed when it didn't. I found a reference for Scons and worked out how it was being used as a wrapper by the python script to do something after compilation even though the python script executed before compilation (that makes your head hurt). That lead me to the "arm-none-eabi-objcopy [option(s)] in-file [out-file]" that was being executed after compilation and telling us it was having a problem, just not telling us what the problem was or even where this instruction was comming from. I had already put print statements into the python script to report the file paths so I new they were being constructed and used in this "arm-none-eabi-objcopy [option(s)] in-file [out-file]" command. Knowing my source file actually existed I was able to type the command into a terminal with the reported file paths and only then did it show the input file path had only been read up to the space in the name of the containing folder because it showed the command as seen by "arm-none-eabi-objcopy [option(s)] in-file [out-file]" in the terminal output with everything after the space missing. So it obviously couldn't resolve the input path file name.

This is more of a problem on Linux machines which have always been much more tolerant of folder name variation than windows. I don't have the facility (well I do but I don't want to risk screwing up my printer) to test the effect of not removing the piece of code or setting/not setting the offset. It may well depend on weather you update just by using the boot-loader or by using flashing software from a PC. But at least it will compile the way Tronxy intended.
